javalib/src/main/scala/java/util/IdentityHashMap.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.util
+
+import java.{util => ju}
+
+import IdentityHashMap.IdentityBox
+
+/**
+ * Does not support Double.NaN keys.
+ */
+class IdentityHashMap[K, V](inner: HashMap[IdentityBox[K], V])
+    extends AbstractMap[K, V] with Map[K, V] with Serializable with Cloneable {
+
+  def this(expectedMaxSize: Int) =
+    this(new HashMap[IdentityBox[K], V](expectedMaxSize, HashMap.DEFAULT_LOAD_FACTOR))
+
+  def this() =
+    this(HashMap.DEFAULT_INITIAL_CAPACITY)
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = {
+    this(initialMap.size())
+    putAll(initialMap)
+  }
+
+  override def clear(): Unit = inner.clear()
+
+  override def clone(): AnyRef =
+    new IdentityHashMap(inner.clone().asInstanceOf[HashMap[IdentityBox[K], V]])
+
+  override def containsKey(key: Any): Boolean =
+    inner.containsKey(IdentityBox(key))
+
+  override def containsValue(value: Any): Boolean = inner.containsValue(value)
+
+  override def get(key: Any): V = inner.get(IdentityBox(key))
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  override def put(key: K, value: V): V = inner.put(IdentityBox(key), value)
+
+  override def remove(key: Any): V = inner.remove(IdentityBox(key))
+
+  override def size(): Int = inner.size
+
+  override def values(): Collection[V] = inner.values()
+
+  override def entrySet(): Set[Map.Entry[K, V]] = new EntrySet
+
+  // views
+  private final class EntrySet extends AbstractSet[Map.Entry[K, V]] {
+
+    def iterator(): Iterator[Map.Entry[K, V]] = {
+      new ju.Iterator[Map.Entry[K, V]] {
+        val iter = inner.entrySetIterator()
+
+        def hasNext(): Boolean = iter.hasNext()
+
+        def next(): Map.Entry[K, V] = {
+          val e = iter.next()
+          new MapEntry(e)
+        }
+
+        def remove(): Unit = iter.remove()
+      }
+    }
+
+    def size(): Int = inner.size()
+
+    override def contains(value: Any): Boolean =
+      value match {
+        case value: Map.Entry[_, _] =>
+          val mapVal = inner.get(IdentityBox(value.getKey()))
+          mapVal == value.getValue()
+        case _ => false
+      }
+
+    override def remove(value: Any): Boolean =
+      value match {
+        case value: Map.Entry[_, _] =>
+          val mapVal = inner.remove(IdentityBox(value.getKey()))
+          mapVal == value.getValue()
+        case _ => false
+      }
+
+    override def clear(): Unit =
+      inner.clear()
+  }
+
+  private final class MapEntry(entry: Map.Entry[IdentityBox[K], V]) extends Map.Entry[K, V] {
+
+    override def equals(other: Any): Boolean =
+      other match {
+        case other: Map.Entry[_, _] =>
+          (entry.getKey() eq IdentityBox(other.getKey())) &&
```

Comment:
You're comparing for reference equality with a freshly allocated `IdentityBox`. That will always return `false`. You should use `eq` between `this.getKey()` and `other.getKey()` instead.

Reply:
I thinking to use this instead so I can avoid repeating the `asInstanceOf[AnyRef]` in the `IdentityBox` code but it does allocate. What do you think?
```
entry.getKey().equals(IdentityBox(other.getKey())) &&
          Objects.equals(this.getValue(), other.getValue())
```